File: Lessons/09_MaximumSliceProblem_MaxProfit.rb
# you can write to stdout for debugging purposes, e.g.
# puts "this is a debug message"

def solution(a)
  # write your code in Ruby 2.2
  n = a.length 
  min = a[0]
  maxDiff = -1 * (10 ** 10)

  1.upto(n - 1) do |i|
    min = a[i] if min > a[i]
    diff = a[i] - min  
    maxDiff = diff if maxDiff < diff 
  end
  return [maxDiff, 0].max
end

File: Python/09_MaximumSliceProblem_MaxProfit.py
# you can write to stdout for debugging purposes, e.g.
# print("this is a debug message")

def solution(A):
    # write your code in Python 3.6
    if len(A) <= 1: return 0
    minimum = A[0]
    ans = 0

    for i in range(1, len(A)):
        tmp = A[i] - minimum 
        minimum = min(A[i], minimum)
        ans = max(tmp, ans)
    
    return ans
